_posts/2022-05-17-calldata-reencode-size-check-bug.md

```diff
@@ -0,0 +1,39 @@
+---
+layout: post
+published: true
+title: 'Size check on ABI-reencoding nested calldata arrays Bug'
+date: '2022-05-17'
+author: Solidity Team
+category: Security Alerts
+---
+
+On April 7, 2022, a bug in the Solidity code generator was reported by
+John Toman of the Certora development team. Certora's bug disclosure post can be found [here](https://medium.com/certora/incorrect-calldata-validation-in-inter-contract-communication-certora-bug-disclosure-8a467ce92ca3)
+
+The bug is fixed with [Solidity version 0.8.14](https://github.com/ethereum/solidity/releases/tag/v0.8.14)
+released on May 17, 2022. The bug was first introduced in Solidity version 0.5.8.
+
+We assigned the bug a severity of "very low".
+
+## Which Contracts are Affected?
+
+You might be affected if you pass a nested array directly to another external
+function call or use ``abi.encode`` on it.
+
+If calldata is malformed in a certain way, the call might not revert as it should.
+Instead, data with zeros is passed on to the called contract or to the ``abi.encode`` function.
+
+## Technical Details
+
+Calldata validation for nested dynamic types is deferred until the first access
+to the nested values. Such an access may for example be a copy to memory or an
+index or member access to the outer type. While in most such accesses calldata
+validation correctly checks that the data area of the nested array is completely
+contained in the passed calldata (i.e. in the range ``[0, calldatasize()]``), this
+check may not be performed, when ABI encoding such nested types again directly from calldata.
+
+For instance, this can happen, if a value in calldata with a nested dynamic array is
+passed to an external call, used in ``abi.encode`` or emitted as event.
+In such cases, if the data area of the nested array extends beyond ``calldatasize()``,
+ABI encoding it did not revert, but continued reading values from beyond
+``calldatasize()`` (i.e. zero values).
```

Comment:
As it is now, this is still quite abstract. At some point you listed examples where this could actually go wrong, maybe we should add something like that?

Reply:
Maybe we should also add an example of badly encoded input from one of the test cases in https://github.com/ethereum/solidity/pull/12907?

Reply:
Not sure... then we also need to explain the format of that data, etc., and it'd become quite verbose...

Reply:
I think it wouldn't really be that much text to just mention how offset and head/tail work. But fair enough, we're pretty close to the release so maybe not the best moment to be doing bigger changes.